### PR TITLE
#12325 fix TLS client endpoint parser handling of bindAddress=

### LINKED ISSUE #12325: TLS client endpoint parser mishandles bindAddress
**Describe the incorrect behavior you saw**

Creating a TLS client endpoint from a string, and including a `bindAddress=ADDR` value (e.g. `tls:github.com:443:bindAddress=1.2.3.4` will cause a crash during connection.

The `twisted.internet.endpoints._parseClientTLS` function passes the `ADDR` into the `HostnameEndpoint()` constructor as a *string*, whereas the function is expecting a tuple of `(ADDR, PORT)`. This doesn't trigger an error during Endpoint construction, however actually trying to connect with the endpoint (`ep.connect(factory)`) gets an error as it attempts to dereference the first element of the tuple and examine it as a hostname.

The following test program ought to connect to Github's TLS endpoint and say hello (note that `bindAddress=`, without a hostname, is a legitimate way to bind to the "ANY" interface, and avoids the need to e.g. call `ifconfig` to figure out a suitable host for binding first.. `bindAddress=0.0.0.0` should work the same way).

```python
from twisted.internet import endpoints, task, protocol, defer

class P(protocol.Protocol):
    def connectionMade(self):
        print("connectionMade")
        self.transport.write(b"GET / HTTP/1.0\n\n")
        self.transport.loseConnection()
    def connectionLost(self, reason):
        self.factory.d.callback("done")
    def dataReceived(self, data):
        print("dataReceived", data)

@defer.inlineCallbacks
def go(reactor):
    #ep = endpoints.clientFromString(reactor, 'tcp:github.com:443:bindAddress=')
    #print(ep._bindAddress)
    ep = endpoints.clientFromString(reactor, 'tls:github.com:443:bindAddress=')
    #print(ep._wrappedEndpoint._bindAddress)
    client_f = protocol.Factory.forProtocol(P)
    client_f.d = defer.Deferred()
    c = yield ep.connect(client_f)
    yield client_f.d
task.react(go)
```

yields the following error:

```
% python3 t.py
main function encountered error
Traceback (most recent call last):
  File "/Users/warner/stuff/python/twisted/src/twisted/internet/defer.py", line 2013, in _inlineCallbacks
    result = context.run(
  File "/Users/warner/stuff/python/twisted/src/twisted/python/failure.py", line 467, in throwExceptionIntoGenerator
    return g.throw(self.value.with_traceback(self.tb))
  File "/Users/warner/stuff/python/twisted/t.py", line 21, in go
    c = yield ep.connect(client_f)
  File "/Users/warner/stuff/python/twisted/src/twisted/internet/endpoints.py", line 600, in connect
    self._reactor.connectTCP(
  File "/Users/warner/stuff/python/twisted/src/twisted/internet/posixbase.py", line 368, in connectTCP
    c.connect()
  File "/Users/warner/stuff/python/twisted/src/twisted/internet/base.py", line 1266, in connect
    self.transport: Optional[Client] = self._makeTransport()
  File "/Users/warner/stuff/python/twisted/src/twisted/internet/tcp.py", line 1529, in _makeTransport
    return Client(self.host, self.port, self.bindAddress, self, self.reactor)
  File "/Users/warner/stuff/python/twisted/src/twisted/internet/tcp.py", line 730, in __init__
    if abstract.isIPv6Address(bindAddress[0]):
builtins.IndexError: string index out of range
```

The `bindAddress[0]` is attempting to extract the host portion of the `(host, port)` tuple, but in this case `bindAddress` is an empty string, so `bindAddress[0]` is an attempt to extract its zeroth character, which fails with IndexError.

If we change the program to use `bindAddress=0.0.0.0`, we get:

```
  File "/Users/warner/stuff/python/twisted/src/twisted/internet/tcp.py", line 1529, in _makeTransport
    return Client(self.host, self.port, self.bindAddress, self, self.reactor)
  File "/Users/warner/stuff/python/twisted/src/twisted/internet/tcp.py", line 734, in __init__
    skt.bind(bindinfo)
builtins.TypeError: bind(): AF_INET address must be tuple, not str
```

because `bindinfo` is a copy of `bindAddress`, and `socket.bind()` expects a tuple.

Changing the program to use the `tcp:` endpoint instead of the `tls:` one yields success:

```
% python3 t.py
connectionMade
```

The problem lies in `_parseClientTLS`. The `bindAddress` is extracted as a string:

https://github.com/twisted/twisted/blob/05cc97afa3c9c00386e5cd0c50e0807e9ef963d4/src/twisted/internet/endpoints.py#L2287-L2291

but then passed into `HostnameEndpoint()` as is:

https://github.com/twisted/twisted/blob/05cc97afa3c9c00386e5cd0c50e0807e9ef963d4/src/twisted/internet/endpoints.py#L2300-L2302

whereas for e.g. `_parseClientTCP`, it converts the string into a `(string, 0)` tuple:

https://github.com/twisted/twisted/blob/05cc97afa3c9c00386e5cd0c50e0807e9ef963d4/src/twisted/internet/endpoints.py#L1880

It looks like the types are not consistent:
* `_parseClientTLS` passes a string `bindAddress` into `HostnameEndpoint`
* [`HostnameEndpoint` is documented](https://github.com/twisted/twisted/blob/05cc97afa3c9c00386e5cd0c50e0807e9ef963d4/src/twisted/internet/endpoints.py#L811-L813) to take a bytestring as `bindAddress`, and after DNS resolution, passes it unmodified into [`TCP6ClientEndpoint`](https://github.com/twisted/twisted/blob/05cc97afa3c9c00386e5cd0c50e0807e9ef963d4/src/twisted/internet/endpoints.py#L975-L980) or [`TCP4ClientEndpoint`](https://github.com/twisted/twisted/blob/05cc97afa3c9c00386e5cd0c50e0807e9ef963d4/src/twisted/internet/endpoints.py#L982-L988)
* [`TCP4ClientEndpoint` is documented](https://github.com/twisted/twisted/blob/05cc97afa3c9c00386e5cd0c50e0807e9ef963d4/src/twisted/internet/endpoints.py#L584-L586) to take `bindAddress` as a *tuple*, as is TCP6ClientEndpoint

I think stronger typechecking might have caught the inconsistency.

`test_endpoints.py` is meant to test this code. However it asserts that the parsed `hostnameEndpoint._bindAddress` is a string, when it should be asserting a tuple.

https://github.com/twisted/twisted/blob/05cc97afa3c9c00386e5cd0c50e0807e9ef963d4/src/twisted/internet/test/test_endpoints.py#L4199

I couldn't find a test that uses `endpointFromtString("tls:...:bindAddress=ADDR")` and then actually tries an `ep.connect()` on it, possibly because generating a suitable TLS server cert is too much of a hassle, and also because it's not obvious that you can use `bindAddress=` or `bindAddress=0.0.0.0` to avoid also needing to probe for a safe address to bind to. The `test_tls` case uses `endpoint.connect()`, but I think with mocked out values which means it doesn't make it all the way to `_BaseTCPClient` and the call to `skt.bind()`, where the error would arise.

My recommended fix is:
* add an `assert(isinstance(bindaddress, (tuple, type(None))))` to _BaseTCPClient.__init__`, `TCP4ClientEndpoint`, and `TCP6ClientEndpoint`
* change the specification of `HostnameEndpoint` to expect a `tuple` for the `bindAddress`
* change `_parseClientTLS()` to call `HostnameEndpoint()` with `(bindAddress, 0)` instead of `bindAddress`
* update test_endpoints.py to check tuple-ness of parsed bindAddress

I'd really like to incorporate a proper `ep.connect()` test, but I didn't see an obvious way to do so.. maybe someone more familiar with the code can come up with a good approach.

I'll attach a PR with the fix and an update to the unit tests.

**Testing environment**

MacOS and Linux. Twisted 24.7.0 and current trunk. Default reactor.

### PR DESCRIPTION
**Pull request by warner**
## Scope and purpose

Fixes `_parseClientTLS` to supply `HostnameEndpoint()` with `bindAddress=` as a tuple, not a string, to avoid a subsequent crash during `endpoint.connect()`. Updates the docstring on `HostnameEndpoint()` to point out that it requires a tuple. Adds (perhaps too many) assertions of tuple-ness to catch similar problems earlier next time. Updates unit tests.

Fixes #12325

### CODE REVIEW

**Comment by warner** [comment]
Huh, so clearly those asserts are too aggressive, since lots of other tests started failing. Or, those other tests are similarly confused about what `bindAddress` ought to be generated. Or *I'm* confused about the same.

I'll dig deeper.

**Comment by warner** [comment]
Results of digging deeper are in the issue: https://github.com/twisted/twisted/issues/12325#issuecomment-2381471729

**Comment by adiroiban** [comment]
Hi Brian. Thanks for this PR that is nicely put together.

Is this ready for review ?

**Comment by warner** [comment]
yes please

Things I'm not sure of:
* str vs bytes? should `ep._bindAddress` always hold one or the other? I think the status quo is to hold whatever the caller provided, which seems non-ideal
* am I testing enough? I'm not satisfied that my tests don't capture the original crash (`ep.connect()`), do you know of a test I could copy to build something more end-to-end?
* are the assertions I added too much? they might not be sufficiently idiomatic. I've been working in JavaScript and Rust for a while and I've gotten paranoid about making type mistakes in untyped languages, so I might be overrotating on the asserts.

**Comment by adiroiban** [comment]
just to trigger the review automation


please review

File: src/twisted/internet/test/test_endpoints.py
Comment (adiroiban) [inline src/twisted/internet/test/test_endpoints.py]: Do we need `nativeString` here ?




```suggestion
        self.assertEqual(hostnameEndpoint._bindAddress, "127.0.0.1", 0))
```

I have checked the implementation of nativeString and I don't see why we need it here.

https://github.com/twisted/twisted/blob/075191d3fd03193e48247fff39e108c784962da2/src/twisted/python/compat.py#L329-L344

File: src/twisted/newsfragments/12325.bugfix
Comment (adiroiban) [inline src/twisted/newsfragments/12325.bugfix]: I a coding standard for the release note, we try to mention the public API affected by this change.

I guess that this is `clientFromString`, but I am not sure.

Also, we have tests, so we are sure it works :) I don't think that we need the "should" part 


```suggestion
twisted.internet.endpoints.clientFromString for TLS endpoint with "bindAddress=" no longer crashes during connect. The HostnameEndpoint() "bindAddress" argument now accepts either hostnames or (host,port) tuples.
```

File: src/twisted/internet/endpoints.py
Comment (adiroiban) [inline src/twisted/internet/endpoints.py]: Can we have the documentation for the types for the `(host, port)` tupple ?

I guess `port` is `int` ... but what about `host` ?

From `def test_tuple(self):` it looks like this is str.

------

Also, what does `local address` mean ?
I think that I know what it means ... but it's not clear if FQDN or hostnames are accepted.

Do we support things other than IPv4 and IPv6 addresses ? 

But maybe this is understood from the context.

File: src/twisted/internet/test/test_endpoints.py
Comment (adiroiban) [inline src/twisted/internet/test/test_endpoints.py]: Just a feedback to your comment

Since `_bindAddresss` is private, I think that it would help to convert it to `str` if needed

I hope that this should simplify the internal code

File: src/twisted/internet/endpoints.py
Comment (adiroiban) [inline src/twisted/internet/endpoints.py]: was this needed by `mypy` ?

If not, I think that the assertions are too much.

With mypy, we should no longer worry about asserting for types.

**Review by adiroiban — APPROVED** [review]
Changes looks good.

I left more comments inline

I think that the change to `test_tls` for endpoint is ok

I see that in the initial GitHub issues you also mentioned this use case... 

```
ep = endpoints.clientFromString(reactor, 'tls:github.com:443:bindAddress=')
```

Is this a real world use case?

If yes, it might help to have a tests similar to `test_tls` ... but simplified , and also documentation for it.

--------

It might help to get a review from @glyph 

I don't really understand how clientFromString should be used .
I had a chat with Glyph and he told me that I am not using the API the was it was designed

File: src/twisted/internet/endpoints.py
Comment (warner) [reply src/twisted/internet/endpoints.py]: Ok, I'll remove it, I added it to try to surface errors earlier.

I don't know how comprehensive mypy's checks are (I kinda dropped out of python work about the same time it was starting up). I do know that this bug exists in current trunk, which suggests that mypy didn't catch it. Maybe that means we need more type annotations? I don't think `HostnameEndpoint.__init__`'s arguments have any annotations (I'm guessing the docstring doesn't count).

Comment (adiroiban) [reply src/twisted/internet/endpoints.py]: The docstring doesn't count for type checking.

The type description from docstring should be moved to python type annotation.
From there, the type annotation for each argument will be automatically reflected in the API doc generation.

File: src/twisted/internet/test/test_endpoints.py
Comment (warner) [reply src/twisted/internet/test/test_endpoints.py]: will do, I'll make it accept (and test) all of:
* None
* bytes
* str
* tuple(bytes, int)
* tuple(str, int)

File: src/twisted/internet/endpoints.py
Comment (warner) [reply src/twisted/internet/endpoints.py]: How does this read?:
```

        @param bindAddress: The client socket normally uses whatever
            local interface (eth0, en0, lo, etc) is best suited for the
            target address, and a randomly-assigned port. This argument
            allows that local address/port to be overridden. Providing
            just an address (as a str) will bind the client socket to
            whichever interface is assigned that address. Providing a
            tuple of (str, int) will bind it to both an interface and a
            specific local port. To bind the port, but leave the
            interface unbound, use a tuple of ("", port), or ("0.0.0.0",
            port) for IPv4, or ("::0", port) for IPv6. To leave both
            interface and port unbound, just use None.
        @type bindAddress: L{str}, L{tuple}, or None
```

File: src/twisted/newsfragments/12325.bugfix
Comment (warner) [reply src/twisted/newsfragments/12325.bugfix]: how's this:
```
twisted.internet.endpoints.clientFromString for TLS endpoints with "bindAddress="  no longer crashes during 
connect. twisted.internet.endpoints.HostnameEndpoint() no longer crashes when given a bindAddress= 
argument that is just a string, and that argument now accepts either address strings or (address, port) tuples.
```

File: src/twisted/internet/test/test_endpoints.py
Comment (warner) [reply src/twisted/internet/test/test_endpoints.py]: I think we don't, I'll remove it and the other nearby uses.

**Comment by warner** [comment]
how do I request a review from @glyph ? lemme see what this does:

please review

File: src/twisted/internet/endpoints.py
Comment (adiroiban) [reply src/twisted/internet/endpoints.py]: Thanks. Looks good.

**Comment by adiroiban** [comment]
To request a review you can use the GitHub UI 

image


We usually request a review from the team... so any team member can review and approve the change

-----

Mentioning another team member can also do the job.

But I don't know if Glyph will have time for the review.

So, as long as the changes from the PR looks good, we can have them merge

We can follow with more changes at any time

File: src/twisted/newsfragments/12325.bugfix
Comment (adiroiban) [reply src/twisted/newsfragments/12325.bugfix]: Looks good. Thanks!

**Review by adiroiban — APPROVED** [review]
Changes looks good. Many thanks!

@warner feel free to merge this PR if you are also happy with the changes.

**Comment by glyph** [comment]
Feel free to @-mention me whenever you want me to look at something, but let's not request reviews from individual people; the robot requests from twisted-contributors on purpose, we already have enough process bottlenecks :)

File: src/twisted/internet/test/test_endpoints.py
Comment (glyph) [inline src/twisted/internet/test/test_endpoints.py:2735]: It would be nice to add a quick docstring to each of these explaining what it is trying to test, but, not a blocker.

File: src/twisted/internet/endpoints.py
Comment (glyph) [inline src/twisted/internet/endpoints.py]: `nativeString` is a holdover from python2 compatibility, we should not really be using it in new code.
```suggestion
                bindAddress = (bindAddress[0].decode(), bindAddress[1])
```

File: src/twisted/internet/endpoints.py
Comment (glyph) [inline src/twisted/internet/endpoints.py:2317]: This precludes binding a client port with a string, which we should probably address with a follow-up somehow.

File: src/twisted/internet/test/test_endpoints.py
Comment (warner) [reply src/twisted/internet/test/test_endpoints.py:2735]: yeah, I figured the docstring on the class covered it well enough

File: src/twisted/internet/endpoints.py
Comment (warner) [reply src/twisted/internet/endpoints.py:2317]: I don't understand. The `:`-joined format for TCP endpoints (which is embedded in the format for TLS client endpoints) doesn't have room for a port, only the binding address as a string/bytes. So previously we were effectively always binding the socket to (string, 0), and this makes that explicit.

(Looking at it again, I think my changes to `HostnameEndpoint.__init__` mean that this change to `_parseClientTLS` is unnecessary. If we didn't do either, then we hit the crash that I observed. I made the `_parseClientTLS` change first, which fixed my most immedaite problem, but really the `HostnameEndpoint` fix is the most complete one. I think I'll leave them both in place, though).

To "[bind] a client port with a string", you use something like `tls:..:endpoint=\"example.org:443:bindAddress=1.2.3.4\"`, and the `1.2.3.4` string is glued into a `("1.2.3.4", 0)` tuple and given to `HostnameEndpoint`.

What am I missing?

**Comment by warner** [comment]
> Feel free to @-mention me whenever you want me to look at something, but let's not request reviews from individual people; the robot requests from twisted-contributors on purpose, we already have enough process bottlenecks :)

Aye, makes sense. I tagged you just because in https://github.com/twisted/twisted/pull/12326#pullrequestreview-2341602050 @adiroiban said:

> I don't really understand how clientFromString should be used . I had a chat with Glyph and he told me that I am not using the API the was it was designed

so I figured we were both missing something, and should follow your lead.

**Comment by warner** [comment]
Ok, assuming CI passes, I think this can be merged. @adiroiban what is the magic bot command to trigger that? (i've really been away too long)

(I don't have the privs to merge myself, in fact now that I look at the "Request Review" corner of this github page, I don't have the privs to request a review either)

**Comment by adiroiban** [comment]
I have merged it.

Brian, I sent an invitation for the twisted-contributors team

this will give you merge access

File: src/twisted/internet/endpoints.py
Comment (glyph) [reply src/twisted/internet/endpoints.py:2317]: > So previously we were effectively always binding the socket to (string, 0), and this makes that explicit

Yes, sorry, I didn't mean to say that this made things *worse*, just that it makes the restriction explicit, and having *noted* that restriction, we should lift it at some point.  We probably want to do it with a separate `bindPort` parameter or something, rather than escaping a colon in here.

**Pull request merged**
